internal/integrations/alerting/alerter.go

```diff
@@ -0,0 +1,152 @@
+package alerting
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/hatchet-dev/hatchet/internal/encryption"
+	"github.com/hatchet-dev/hatchet/internal/integrations/alerting/alerttypes"
+	"github.com/hatchet-dev/hatchet/internal/integrations/email"
+	"github.com/hatchet-dev/hatchet/internal/repository"
+	"github.com/hatchet-dev/hatchet/internal/repository/prisma/db"
+	"github.com/hatchet-dev/hatchet/internal/repository/prisma/sqlchelpers"
+
+	"github.com/hatchet-dev/timediff"
+)
+
+type TenantAlertManager struct {
+	repo      repository.EngineRepository
+	enc       encryption.EncryptionService
+	serverURL string
+	email     email.EmailService
+}
+
+func New(repo repository.EngineRepository, e encryption.EncryptionService, serverURL string, email email.EmailService) *TenantAlertManager {
+	return &TenantAlertManager{repo, e, serverURL, email}
+}
+
+func (t *TenantAlertManager) HandleAlert(tenantId string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// read in the tenant alerting settings and determine if we should alert
+	tenantAlerting, err := t.repo.TenantAlertingSettings().GetTenantAlertingSettings(ctx, tenantId)
+
+	if err != nil {
+		return err
+	}
+
+	lastAlertedAt := tenantAlerting.Settings.LastAlertedAt.Time.UTC()
+	maxFrequency, err := time.ParseDuration(tenantAlerting.Settings.MaxFrequency)
+
+	if err != nil {
+		return err
+	}
+
+	if lastAlertedAt.IsZero() || time.Since(lastAlertedAt) > maxFrequency {
+		// update the lastAlertedAt
+		now := time.Now().UTC()
+
+		err = t.repo.TenantAlertingSettings().UpdateTenantAlertingSettings(ctx, tenantId, &repository.UpdateTenantAlertingSettingsOpts{
+			LastAlertedAt: &now,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		return t.sendAlert(ctx, tenantAlerting, lastAlertedAt)
+	}
+
+	return nil
+}
+
+func (t *TenantAlertManager) SendAlert(tenantId string, prevLastAlertedAt time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// read in the tenant alerting settings and determine if we should alert
+	tenantAlerting, err := t.repo.TenantAlertingSettings().GetTenantAlertingSettings(ctx, tenantId)
+
+	if err != nil {
+		return err
+	}
+
+	return t.sendAlert(ctx, tenantAlerting, prevLastAlertedAt)
+}
+
+func (t *TenantAlertManager) sendAlert(ctx context.Context, tenantAlerting *repository.GetTenantAlertingSettingsResponse, prevLastAlertedAt time.Time) error {
+	// read in all failed workflow runs since the last alerted time, ordered by the most recent runs first
+	statuses := []db.WorkflowRunStatus{
+		db.WorkflowRunStatusFailed,
+	}
+
+	limit := 5
+
+	tenantId := sqlchelpers.UUIDToStr(tenantAlerting.Settings.TenantId)
+
+	failedWorkflowRuns, err := t.repo.WorkflowRun().ListWorkflowRuns(
```

Comment:
we'll likely want to filter by workflow/namespace?

Reply:
eventually yes, might be a more general alerting setting that we expose to the user. right now alerts are opt-in and users are typically creating multiple tenants to split up dev/prod.

Reply:
i think we'll want to put this on the alert group object since users will likely want different "routing" for different workflows